internal/ykclient/client.go

```diff
@@ -36,50 +39,88 @@ func NewClient(httpProto string, ykHost string, ykPort int, repo *repository.Rep
-	resp, err := http.Get(streamURL)
+
+	evCounts, ok := ctx.Value(config.EventCounts).(config.EventTypeCounts)
+	if !ok || (evCounts == nil) {
+		fmt.Fprintf(os.Stderr, "could not get eventCounts map from context\n")
+		return
+	}
+
+	go func() {
+		fmt.Println("Starting YuniKorn event stream client")
+		c.FetchEventStream(ctx, streamURL, evCounts)
+	}()
+}
+
+func (c *Client) FetchEventStream(ctx context.Context, streamURL string, evCounts config.EventTypeCounts) {
+	ctx, cancel := context.WithCancel(ctx)
+
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, streamURL, bytes.NewBuffer([]byte{}))
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Error: could not create new HTTP request: %v\n", err)
+		cancel()
+		return
+	}
+
+	client := &http.Client{}
+
+	resp, err := client.Do(req)
 	if err != nil {
 		fmt.Fprintf(os.Stderr, "could not request from %s: %v", streamURL, err)
+		cancel()
+		return
 	}
+	defer func() {
+		err := resp.Body.Close()
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error: could not close body of event stream connection: %v\n", err)
+		}
+	}()
 
 	reader := bufio.NewReader(resp.Body)
-	go func() {
-		fmt.Println("Starting YuniKorn event stream client")
-		for {
-			select {
-			case <-ctx.Done():
-				// TODO: add logging here to indicate that the client is shutting down
-				return
-			default:
-				line, err := reader.ReadBytes('\n')
-				if err != nil {
-					fmt.Fprintf(os.Stderr, "Error: could not read from http stream: %v", err)
-					break
-				}
-
-				ev := si.EventRecord{}
-				err = json.Unmarshal(line, &ev)
-				if err != nil {
-					fmt.Fprintf(os.Stderr, "could not unmarshal event from stream: %v\n", err)
-					break
-				}
-				// TODO: This is Okayish for small number of events, but for large number of events this will be a bottleneck
-				// We should consider using a channel? or a pool of workers? or a different queuing system ? to handle events.
-				c.handleEvent(ctx, &ev)
-
-				if ev.Type == si.EventRecord_APP {
-					fmt.Printf("Application\n")
-					fmt.Printf("---------\n")
-					fmt.Printf("Type         : %s\n", si.EventRecord_Type_name[int32(ev.Type)])
-					fmt.Printf("ObjectId     : %s\n", ev.ObjectID)
-					fmt.Printf("Message      : %s\n", ev.Message)
-					fmt.Printf("Change Type  : %s\n", ev.EventChangeType)
-					fmt.Printf("Change Detail: %s\n", ev.EventChangeDetail)
-					fmt.Printf("Reference ID:  %s\n", ev.ReferenceID)
-					fmt.Printf("Resource    : %+v\n", ev.Resource)
-				}
+
+	for {
+		select {
+		case <-ctx.Done():
+			// TODO: add logging here to indicate that the client is shutting down
+			cancel()
+			return
+		default:
+			line, err := reader.ReadBytes('\n')
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "Error: could not read from http stream: %v", err)
+				break
+			}
+
+			ev := si.EventRecord{}
+			err = json.Unmarshal(line, &ev)
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "could not unmarshal event from stream: %v\n", err)
+				break
+			}
```

Comment:
Can we capture this in a ticket?

Reply:
This is now described in ticket https://github.com/G-Research/yunikorn-history-server/issues/34